tetrominos_learning/tetrominos_learning.py

```diff
@@ -0,0 +1,291 @@
+        Dgate(tf.clip_by_value(dr2[l], -disp_clip, disp_clip), dphi2[l]) | q[1]
+        Kgate(tf.clip_by_value(kappa1[l], -kerr_clip, kerr_clip)) | q[0]
+        Kgate(tf.clip_by_value(kappa2[l], -kerr_clip, kerr_clip)) | q[1]
+
+
+# StrawberryFields quantum simulator
+engine, q = sf.Engine(num_subsystems=modes)
+
+# Definition of the CV quantum network
+with engine:
+    # State preparation
+    Dgate(disps_alpha) | q[0]
+    Dgate(disps_beta) | q[1]
+    # Sequence of variational layers
+    for i in range(depth):
+        layer(i)
+
+# Symbolic evaluation of the output state
+state = engine.run('tf', cutoff_dim=cutoff, eval=False, batch_size=num_images)
+ket = state.ket()
+
+# Projection on the subspace of up to im_dim-1 photons for each mode.
+ket_reduced = ket[:, :im_dim, :im_dim]
+norm = tf.sqrt(tf.abs(tf.reduce_sum(tf.conj(ket_reduced) * ket_reduced, axis=[1, 2])))
+norm_inv = tf.expand_dims(
+    tf.expand_dims(tf.cast(1.0 / norm, dtype=tf.complex64), axis=[1]), axis=[1]
+)
+ket_processed = ket_reduced * norm_inv
+
+# ====================================================
+# 		 Definition of the loss function
+# ====================================================
+
+# Target images
+data_states = tf.placeholder(tf.complex64, shape=[num_images, im_dim, im_dim])
+
+# Overlaps with target images
+overlaps = tf.abs(tf.reduce_sum(tf.conj(ket_processed) * data_states, axis=[1, 2])) ** 2
+
+# Overlap cost function
+overlap_cost = tf.reduce_mean((overlaps - 1) ** 2)
+
+# State norm cost function
+norm_cost = tf.reduce_sum(
+    (tf.abs(tf.reduce_sum(tf.conj(ket) * ket, axis=[1, 2])) ** 2 - 1) ** 2
+)
+
+cost = overlap_cost + norm_weight * norm_cost
+
+# ====================================================
+# 	TensorBoard logging of cost functions and images
+# ====================================================
+
+tf.summary.scalar('Cost', cost)
+tf.summary.scalar('Norm cost', norm_cost)
+tf.summary.scalar('Overlap cost', overlap_cost)
+
+# Output images with and without subspace projection.
+images_out = tf.abs(ket_processed)
+images_out_big =tf.abs(ket)
+tf.summary.image('image_out', tf.expand_dims(images_out, axis=3), max_outputs=num_images)
+tf.summary.image('image_out_big', tf.expand_dims(images_out_big, axis=3), max_outputs=num_images)
+
+# TensorBoard writer and summary
+writer = tf.summary.FileWriter(board_string)
+merge = tf.summary.merge_all()
+
+
+# ====================================================
+# 					Training
+# ====================================================
+
+# Optimization algorithm (Adam optimizer)
+optim = tf.train.AdamOptimizer()
+training = optim.minimize(cost)
+
+print('Graph building time: {:3f}'.format(time.time() - init_time))
+
+# TensorFlow session
+with tf.Session() as session:
+    session.run(tf.global_variables_initializer())
+    start_time = time.time()
+
+    for i in range(reps):
+        rep_time = time.time()
+        # make an optimization step
+        [_training] = session.run(
```

Comment:
Does `[_training]` and `[training]` need to be a one element list?

Reply:
No. I can fix it